File: Python/Pattern wise question/DFS/4_max_path_sum.py
https://leetcode.com/problems/binary-tree-maximum-path-sum

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ans = float("-inf")
        def solve(self , root):
            nonlocal ans
            if(not root):
                return 0
            left = solve(self , root.left) + root.val
            right = solve(self , root.right) + root.val
            ans = max(ans , root.val , left , right , left + right - root.val)

            return max(left , right , root.val)
        solve(self , root)
        return ans


# Recursively
def maxPathSum(self, root):
    self.res = -sys.maxsize-1
    self.oneSideSum(root)
    return self.res

# compute one side maximal sum,
# (root+left children, or root+right children),
# root is the included top-most node
def oneSideSum(self, root):
    if not root:
        return 0
    l = max(0, self.oneSideSum(root.left))
    r = max(0, self.oneSideSum(root.right))
    self.res = max(self.res, l+r+root.val)
    return max(l, r)+root.val

File: C++/Pattern wise question/DFS/4_max_path_sum.cpp
https://leetcode.com/problems/binary-tree-maximum-path-sum

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxSum(TreeNode* root, int& ans) {
    /* This function return the Branch Sum......
    So if the node is NULL then it won't have a branch....so the branch sum will be 0.
    */
    //Base Case
    if(root == NULL){
        return 0;
    }

    //Recursive Case
    //BS = Branch Sum
    int leftBS = root->val + maxSum( root->left , ans );
    int rightBS = root->val + maxSum( root->right , ans );

    ans = max({
                ans,            //we may have found the maximum ans already
                root->val,      //may be the current root val is the maximum sum possible
                leftBS,         //may be the answer contain root->val + left branch value
                rightBS,        //may be the answer contain root->val + right branch value
                leftBS + rightBS - root->val   // may be ans conatin left branch + right branch + root->val
                                               // Since the root val is added twice from leftBS and rightBS so we are sunstracting it.
            });

    //Return the max branch Sum
    return max({ leftBS , rightBS , root->val });
}

int maxPathSum(TreeNode* root) {
    int ans = INT_MIN;
    maxSum(root, ans);
    return ans;
}

};
